File: GCTP_CPP/mercator.cpp

#include "mercator.h"
#include "projexception.h"

Mercator::Mercator(): Cylindrical(), m_e(0.0),
m_es(0.0), m_m1(0.0)
{
	setName("Mercator");
	setNumber(MERCAT);
}

Mercator::Mercator(double gctpParams[], int units, long datum, long spheroid):
Cylindrical(gctpParams, units, datum, spheroid), m_e(0.0),
m_es(0.0), m_m1(0.0)
{
		setName("Mercator");
		setNumber(MERCAT);
}

void Mercator::inverse_init() {
	double temp;			/* temporary variable		*/

	temp = m_rMinor / m_rMajor;
	m_es = 1.0 - SQUARE(temp);
	m_e = sqrt(m_es);
	m_m1 = cos(m_centerLat)/(sqrt(1.0 - m_es * sin(m_centerLat) * sin(m_centerLat)));
	m_invInitNeeded = false;

}
void Mercator::inverse(double x, double y, double* lon, double* lat) 
{
	double ts;		/* small t value				*/
	long flag;		/* error flag 					*/
	
	if(m_invInitNeeded)
		inverse_init();
	
	Util::convertCoords(m_unitCode, METER, x, y);

	/* Inverse equations
	-----------------*/
	flag = 0;
	x -= m_falseEasting;
	y -= m_falseNorthing;
	ts = exp(-y/(m_rMajor * m_m1));
	m_latitude = Util::phi2z(m_e,ts,&flag);
	
	if (flag != 0)
		throw(ProjException(flag, "Mercator::inverse()"));

	m_longitude = Util::adjust_lon(m_centerLon + x/(m_rMajor * m_m1));

	Util::convertCoords(RADIAN, DEGREE, m_longitude, m_latitude);

	if(lon)
		*lon = m_longitude;

	if(lat)
		*lat = m_latitude;
}

void Mercator::forward_init() {
	double temp;			/* temporary variable		*/

	temp = m_rMinor / m_rMajor;
	m_es = 1.0 - SQUARE(temp);
	m_e = sqrt(m_es);
	m_m1 = cos(m_centerLat)/(sqrt(1.0 - m_es * sin(m_centerLat) * sin(m_centerLat)));
	m_forInitNeeded = false;
}

void Mercator::forward(double lon, double lat, double* x, double* y) {
	double ts;		/* small t value				*/
	double sinphi;		/* sin value					*/

	if(m_forInitNeeded)
		forward_init();

	Util::convertCoords(DEGREE, RADIAN, lon, lat);

	/* Forward equations
	 -----------------*/
	if (fabs(fabs(lat) - HALF_PI)  <= EPSLN)
		throw(ProjException(52, "Mercator::forward()"));
    
	else
	 {
		sinphi = sin(lat);
		ts = Util::tsfnz(m_e,lat,sinphi);
		m_x_coord = m_falseEasting + m_rMajor * m_m1 * Util::adjust_lon(lon - m_centerLon);
		m_y_coord = m_falseNorthing - m_rMajor * m_m1 * log(ts);
	 }

	 Util::convertCoords(METER, m_unitCode, m_x_coord, m_y_coord);

	 if(x)
		 *x = m_x_coord;
	 if(y)
		 *y = m_y_coord;
}

File: JGCTP/Mercator.java
package JGCTP;

public class Mercator extends Projection
{

protected 	double m_e;

protected 	double m_es;

protected 	double m_m1;

public 	Mercator()
{
    super();
    setName("Mercator");
    setNumber(ProjCode.MERCAT);
    m_e = 0;
    m_es = 0;
    m_m1 = 0;
    
    
}


public 	Mercator(double gctpParams[],  ProjUnit units, ProjDatum dat)
{
    super(gctpParams, units, dat);
    setName("Mercator");
    setNumber(ProjCode.MERCAT);
    m_e = 0;
    m_es = 0;
    m_m1 = 0;
}


protected 	void _init() throws ProjException
{
    double temp;            /* temporary variable       */

    temp = m_sphere.rMinor / m_sphere.rMajor;
    m_es = 1.0 - (temp*temp);
    m_e = Math.sqrt(m_es);
    m_m1 = Math.cos(m_center.lat)/(Math.sqrt(1.0 - m_es * Math.sin(m_center.lat) * Math.sin(m_center.lat)));

}


protected void _forward(GeoPoint p) throws ProjException
{
    double ts;      /* small t value                */
    double sinphi;      /* sin value                    */


    /* Forward equations
     -----------------*/
    if (Math.abs(Math.abs(p.lat) - ((Math.PI)/2))  <= Constants.EPSLN) {
        throw(new ProjException(52, "Mercator._forward()"));
       
    }

    else
     {
        sinphi = Math.sin(p.lat);
        ts = Util.tsfnz(m_e,p.lat,sinphi);
        m_xy.x = m_falseEasting + m_sphere.rMajor * m_m1 * Util.adjust_lon(p.lon - m_center.lon);
        m_xy.y = m_falseNorthing - m_sphere.rMajor * m_m1 * Math.log(ts);
     }

}


protected void _inverse(CoordPoint p) throws ProjException
{
    double ts;      /* small t value                */
   
    /* Inverse equations
    -----------------*/

    p.x -= m_falseEasting;
    p.y -= m_falseNorthing;
    ts = Math.exp(-p.y/(m_sphere.rMajor * m_m1));
    
    try
    {
        m_lonLat.lat = Util.phi2z(m_e,ts);
    }
    
    catch(ProjException e)
    {
        e.addFunc("Mercator._inverse()");
        throw(e);
      
    }

    m_lonLat.lon = Util.adjust_lon(m_center.lon + p.x/(m_sphere.rMajor * m_m1));
}



}